library/coretests/tests/iter/adapters/zip.rs

```diff
@@ -328,6 +328,80 @@ fn test_nested_zip_panic_safety() {
     assert_eq!(witness[3].1.load(Ordering::Relaxed), 0);
 }
 
+#[test]
+fn test_default_zip() {
+    // test the un-specialized version, using RepeatN, which does not implement TrustedRandomAccess(NoCoerce)
+    let mut iter = repeat_n(0, 2).zip([1, 2, 3]);
+    assert_eq!(iter.next(), Some((0, 1)));
+    assert_eq!(iter.size_hint(), (1, Some(1)));
+    assert_eq!(iter.next_back(), Some((0, 2)));
+    assert_eq!(iter.next_back(), None);
+}
+
+#[test]
+fn test_zip_evaluation_order() {
+    use std::cell::RefCell;
+    // Check that order of side effects is correct
+
+    let log = RefCell::new(Vec::new());
+    let xs = [0, 1, 2, 3];
+    let ys = [10, 20];
+    let mut iter1 = xs
+        .iter()
+        .map(|x| {
+            log.borrow_mut().push(0);
+            x
+        })
+        .zip(ys.iter().map(|x| {
+            log.borrow_mut().push(1);
+            x
+        }));
+    iter1.next();
+    iter1.next_back();
+    assert_eq!(&log.take(), &[0, 1, 0, 0, 0, 1]);
+
+    let mut iter2 = ys
+        .iter()
+        .map(|x| {
+            log.borrow_mut().push(0);
+            x
+        })
+        .zip(xs.iter().map(|x| {
+            log.borrow_mut().push(1);
+            x
+        }));
+    iter2.next();
+    iter2.next_back();
+    assert_eq!(&log.take(), &[0, 1, 1, 1, 0, 1]);
+
+    let mut iter3 = repeat_n(0, 4)
+        .map(|x| {
+            log.borrow_mut().push(0);
+            x
+        })
+        .zip(ys.iter().map(|x| {
+            log.borrow_mut().push(1);
+            x
+        }));
+    iter3.next();
+    iter3.next_back();
+    assert_eq!(&log.take(), &[0, 1, 0, 0, 0, 1]);
+
+    let mut iter4 = ys
+        .iter()
+        .map(|x| {
+            log.borrow_mut().push(0);
+            x
+        })
+        .zip(repeat_n(0, 4).map(|x| {
+            log.borrow_mut().push(1);
+            x
+        }));
+    iter4.next();
+    iter4.next_back();
+    assert_eq!(&log.take(), &[0, 1, 1, 1, 0, 1]);
+}
+
```

Comment:
This tests for the ordering of calls to the constituent iterators of a Zip, but I don't think that ordering is specified, nor do I think it should be...